openfeature/multi/multiprovider.go

```diff
@@ -350,106 +355,137 @@ func (p *Provider) ObjectEvaluation(ctx context.Context, flag string, defaultVal
-			if p.updateProviderStateFromEvent(e) {
-				p.outboundEvents <- e.Event
-				l.LogAttrs(context.Background(), slog.LevelDebug, "forwarded state update event")
-			} else {
-				l.LogAttrs(context.Background(), slog.LevelDebug, "total state not updated, inbound event will not be emitted")
-			}
-		}
-	}()
+	if len(handlers) > 0 {
+		p.workerGroup.Add(1)
+		go p.forwardProviderEvents(workerCtx, handlers)
+	} else {
+		// we don't emit any events so we can just close the channel
+		close(p.outboundEvents)
+	}
 
 	p.setStatus(of.ReadyState)
 	p.initialized = true
 	return nil
 }
 
-// startListening is intended to be called on a per-provider basis as a goroutine to listen to events from a provider
-// implementing [of.EventHandler].
-func (p *Provider) startListening(ctx context.Context, name string, h of.EventHandler, wg *sync.WaitGroup) {
-	wg.Add(1)
-	defer wg.Done()
-	for {
-		select {
-		case e := <-h.EventChannel():
-			e.EventMetadata[MetadataProviderName] = name
-			e.EventMetadata[MetadataProviderType] = h.(of.FeatureProvider).Metadata().Name
-			p.inboundEvents <- namedEvent{
-				Event:        e,
-				providerName: name,
+// forwardProviderEvents establishes an event forwarding pipeline that collects events from multiple provider
+// event handlers and forwards them to the multiprovider's outbound event channel. It spawns a goroutine for
+// each provider handler to listen for events, aggregates them through an internal pipe, and selectively forwards
+// events that result in state changes. The function blocks until workerCtx is cancelled or all provider event
+// channels are closed, ensuring proper cleanup by closing the outbound channel when complete.
+func (p *Provider) forwardProviderEvents(workerCtx context.Context, handlers chan namedEventHandler) {
+	defer p.workerGroup.Done()
+	defer close(p.outboundEvents)
+
+	workerLogger := p.logger.With(slog.String("multiprovider-worker", "event-forwarder-worker"))
+	pipe := make(chan namedEvent)
+	var wg sync.WaitGroup
+	for ch := range handlers {
+		wg.Add(1)
+		go func(ctx context.Context, h of.EventHandler, name string, out chan<- namedEvent) {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case e, ok := <-h.EventChannel():
+					if !ok {
+						return
+					}
+					if e.EventMetadata == nil {
+						e.EventMetadata = make(map[string]any)
+					}
+					e.EventMetadata[MetadataProviderName] = name
+					if p, ok := h.(of.FeatureProvider); ok {
+						e.EventMetadata[MetadataProviderType] = p.Metadata().Name
+					}
+					out <- namedEvent{
+						Event:        e,
+						providerName: name,
+					}
+				}
 			}
-		case <-ctx.Done():
-			return
+		}(workerCtx, ch.EventHandler, ch.name, pipe)
+	}
+
+	go func() {
+		wg.Wait()
+		close(pipe)
+	}()
+
+	for e := range pipe {
+		l := workerLogger.With(
+			slog.String(MetadataProviderName, e.providerName),
+			slog.String(MetadataProviderType, e.ProviderName),
```

Comment:
Are these log attrs supposed to have the same values? As in, should MetadataProviderType have value e.ProviderName?

Reply:
I think there’s a naming issue here - I’ve struggled with it too. `providerName` is a user-defined name in the `WithProvider` option in `namedEvent`, while `ProviderName` is the defined name in `of.Event`.